text/0045-docker-proxy.md

```diff
@@ -0,0 +1,148 @@
+
+Starting with Ubuntu 18 (Bionic), `gpg` , invoked through `apt-key adv` , now ignores `http_proxy`
+and `https_proxy` environment variables, and ignores the command line options
+`apt-key adv --keyserver-options http-proxy=<proxy server url>`
+and `https-proxy=<proxy server url>` . The only method remaining is with cumbersome
+apt command-specific configuration file settings in `/etc/apt/apt.conf.d/proxy` .
+
+
+# Goals
+[goals]: #goals
+
+* Have Docker and Docker Compose files and shell scripts that work in explicit proxy,
+  transparent proxy, and non-proxy environments
+* Have a uniform method to access proxy servers across Hyperledger Sawtooth projects
+* Minimize Docker and script clutter with unnecessary tests and settings
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to use the proposed containers, developers and users in proxied environments must set
+these environment variables:
+
+* If you use unsecure HTTP web access, set `http_proxy` . The setting is site-specific for your
+  corporate proxy server. For example,
+  `export http_proxy=http://proxy.foocorp.com:8080`
+* Similarly, if you use secure HTTPS/SSL/TLS web access, set `https_proxy` . The port number or
+  server or both are usually different from `http_proxy`
+* Similarly set `ftp_proxy` if you use FTP (this is not common)
+* Set `no_proxy` to a comma-separated list of internal corporate subnets. For example,
+  `no_proxy=.foocorp.com,10.0.0.0/8,localhost,127.0.0.0/8`
+* Do *not* set `HTTP_PROXY` , `HTTPS_PROXY` , and `NO_PROXY` . Most or all software use the
+  lower case `*_proxy` environment variables and only sometimes use UPPER CASE `*_PROXY` variables
+* Proxy support should be briefly mentioned in the README.md file or similar file, along with the
+  environment variable(s) needed, such as `HTTP_PROXY`
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+All Docker files must be updated to conform as follows:
+* Commands `apt, apt-get, cargo, deb, go, npm, pip3, wget` and others require no changes
+* `curl` requires no changes, but it is not usually installed. It needs to be installed for
+  containers that use it. For example,
+  `RUN apt-get update && apt-get install -yq curl `
+
+* `apt-key` : Instead of using `apt-key adv` to download an apt signing key, use `curl` and
+  `apt-key add -` . For example, instead of the following in a `Dockerfile` (which no longer
+  works):
+
+```
+RUN \
+ apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 \
+ --recv-keys 8AA7AF1F1091A5FD
+```
+
+Use the following, which works in proxy (with http_proxy) and non-proxy environments:
+
+```
+RUN \
+ apt-get update && apt-get install -yq curl && curl -sSL \
+ 'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD’\
+ | apt-key add -
+```
+
+All Docker Compose files must, pass `http_proxy` , `https_proxy` , and `no_proxy` through to each
```

Comment:
This depends on the docker version. Maybe we can mention that Sawtooth assumes user is on latest version and in version 18.03 onwards this explicit passing of env is not required.

Reply:
I didn't know that. I will verify the *_proxy variables are passed to Docker and update the document.

Reply:
I tried a simple test that removed the `*_proxy` variables from a Docker file and the application now fails.  I used my CookieJar application with docker 18.09.6 and docker-compose 1.21.0.  Here is the diff:

```
$ git diff
diff --git a/docker-compose.yaml b/docker-compose.yaml
index 934b1ee..8ac2031 100644
--- a/docker-compose.yaml
+++ b/docker-compose.yaml
@@ -39,10 +39,6 @@ services:
         - http_proxy
         - https_proxy
         - no_proxy
-    environment:
-      - 'http_proxy=${http_proxy}'
-      - 'https_proxy=${https_proxy}'
-      - 'no_proxy=rest-api,validator,${no_proxy}'
     volumes:
       - '.:/project/cookiejar/'
     depends_on:
```

The appliation worked before the change. After rebuilding the docker container with the above change, this is the result:

```
$ sudo docker exec -it cookiejar-cl
ient bash
[sudo] password for dano: 
root@e18112d57b62:/project/cookiejar/pyclient# cookiejar.py bake 100
URL to send to REST API is http://rest-api:8008/batches
[18:47:26 INFO    connectionpool] Starting new HTTP connection (1): proxy-jf.intel.com
[18:47:26 DEBUG   connectionpool] "POST http://rest-api:8008/batches HTTP/1.1" 404 654
Traceback (most recent call last):
  File "/project/cookiejar/pyclient/cookiejar_client.py", line 137, in _send_to_rest_api
    result.status_code, result.reason))
Exception: Error 404: Not Found
```

docker and docker-compose version:
```
 $ docker version
Client:
 Version:           18.09.6
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        481bc77
 Built:             Sat May  4 02:35:57 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.6
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       481bc77
  Built:            Sat May  4 01:59:36 2019
  OS/Arch:          linux/amd64
  Experimental:     false
daniela1-test1 sawtooth-cookiejar (docker) $ docker compose version
docker: 'compose' is not a docker command.
See 'docker --help'
daniela1-test1 sawtooth-cookiejar (docker) $ docker-compose version
docker-compose version 1.21.0, build 5920eb0
docker-py version: 3.2.1
CPython version: 3.6.5
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
```

Reply:
There are two things here:
1. If you pruned all docker images and rebuilt them by removing `args` under the `build` then `config.json` worked as expected. Proxy env variables were passed by docker during build time.
2. These env are also passed during container runtime, hence you saw see the error while executing here. By adding `rest-api` to the `no_proxy` list in `config.json` this issue would be resolved.

We're on same page, how do we handle `no_proxy` is the next question.

Reply:
Sorry, I still do not understand what's being proposed here. This `config.json` file. Does it go in the host? Where? `/etc/docker/config.json` ? Or does it go in each of the Docker containers? If it's the latter, that is much less convenient than just setting environment variables.

Also what environment variables are now automatically passed to containers? Is it `http_proxy` , `https_proxy` , and `no_proxy` ? If these are passed, why is a config.json needed?

If `config.json` (is it `/etc/docker/config.json` ?), can you give me an example?

BTW, I removed the `args` and rebuilt and it still did not work.

Reply:
Ah! I assumed you were aware of these changes from your previous comment. Let me brief it now, `config.json` is a file to be created on the client machine where we want to run the containers. Create this file under `~/.docker/` directory. This file will have the `****_proxy` and other env variables which are to be passed by the docker during build and running the container.

The `config.json` file looks like
```
{
    "proxies":
    {
        "default":
        {
            "httpProxy": "http://proxy-server-domain:<port>",
            "httpsProxy": "http://proxy-server-domain:<port>",
            "noProxy": "*.my-company.com,localhost,rest-api*",
            "hkpProxy": "http://proxy-server-domain:<port>"
        }
    }
}
```

Please let me know if adding this file helped you